### PR TITLE
ci: build node in manylinux docker container

### LINKED ISSUE #359: Build node native modules inside of manylinux container
Make sure they are compatible with as many Linux versions as possible.

Fixing this might also address #204.

### PR DESCRIPTION
**Pull request by wjones127**
Closes #359

TODO:
 * [x] test in a sample of Linux distro docker containers

### CODE REVIEW

**Comment by wjones127** [comment]
✅ Successfully built in CI: https://github.com/lancedb/lancedb/actions/runs/5628619030/job/15252682303

File: ci/build_linux_artifacts.sh [hunk 1]
```diff
@@ -1,76 +1,17 @@
...
-# Others:
-# aarch64-unknown-linux-gnu
-# x86_64-unknown-linux-musl
-# aarch64-unknown-linux-musl
-
-setup_dependencies $TARGET
-install_node $TARGET
-build_node_binary $TARGET
+ARCH=${1:-x86_64}
+
```
Comment (lindt) [inline ci/build_linux_artifacts.sh:2]: is it cool to have scripts without "set -e"?

File: ci/manylinux_node/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ARG ARCH=x86_64
+
+FROM quay.io/pypa/manylinux2014_${ARCH}
+
+ARG ARCH=x86_64
+ARG DOCKER_USER=default_user
+
+# Install static openssl
+COPY install_openssl.sh install_openssl.sh
+RUN bash ./install_openssl.sh ${ARCH} > /dev/null
```
Comment (lindt) [inline ci/manylinux_node/Dockerfile:14]: why not "chmod +x and git commit this"?

File: ci/manylinux_node/install_protobuf.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Installs protobuf compiler. Should be run as root.
+
```
Comment (lindt) [inline ci/manylinux_node/install_protobuf.sh:2]: set -e missing?

File: ci/manylinux_node/prepare_manylinux_node.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+install_rust() {
+    echo "Installing rust..."
+    curl https://sh.rustup.rs -sSf | bash -s -- -y
+    export PATH="$PATH:/root/.cargo/bin"
+}
+
+TARGET=${1:-x86_64} # or aarch64
+
+install_node $TARGET
```
Comment (lindt) [inline ci/manylinux_node/prepare_manylinux_node.sh:21]: parameter $TARGET defined, but not used

File: .github/workflows/npm-publish.yml [hunk 5]
```diff
@@ -177,7 +126,7 @@ jobs:
             node/dist/lancedb-vectordb-win32*.tgz
 
   release:
-    needs: [node, node-macos, node-linux, node-linux-arm, node-windows]
+    needs: [node, node-macos, node-linux, node-windows]
```
[the remark this answers is not in the document]
Comment (wjones127) [reply .github/workflows/npm-publish.yml:129]: This doesn't remove the arm release. It's just that it's combined in GA into the `node-linux` job, which is parameterized by runner now.

File: ci/build_linux_artifacts.sh (hunk 1 above)
Comment (wjones127) [reply ci/build_linux_artifacts.sh:2]: I'll add that to all the scripts.

File: .github/workflows/npm-publish.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (chebbyChefNEQ) [reply .github/workflows/npm-publish.yml:129]: I see, makes sense

**Review by chebbyChefNEQ — COMMENTED** [review]
LGTM, let me do a quick test in a local docker.

**Review by chebbyChefNEQ — APPROVED** [review]
🚀  (tested on buster)

Let's release 0.1.18 after this PR?

```
(base)  ✘ rmeng@rmeng-B650M-DS3H  ~  docker run -v `pwd`:/data -it --entrypoint /bin/bash node:16-buster-slim  
root@cde897d9c10c:/test# npm install vectordb
root@cde897d9c10c:/test# npm i /data/lancedb-vectordb-linux-x64-gnu-0.1.17.tgz 
root@cde897d9c10c:/test# npx ts-node
Need to install the following packages:
  ts-node@10.9.1
Ok to proceed? (y) y
> import {connect} from "vectordb"
undefined
> let conn1 = await connect("/data/workspace/lance/benchmarks/sift")
undefined
> await conn1.tableNames()
[
  'sift1m_base',
  'sift1m_ivf1024_pq16',
  'sift1m_ivf2048_pq16',
  'sift1m_ivf512_pq16'
]
> 
```

**Comment by wjones127** [comment]
✅ Tested on `node:16-buster-slim`, `node:16-bullseye`, `node:20-bookworm`

<details>
<summary>Scripts</summary>

```bash
#!/bin/bash
set -e

# This script assumes that the linux node artifacts have already been built.

# node:16-buster-slim
# node:16-bullseye
# node:20-bookworm
IMAGE=node:20-bookworm
ARCH=$1
VERSION=$2

docker run \
    -v $(pwd)/node:/node -w /node \
    -v $(pwd)/ci/manylinux_node_test:/scripts -w /scripts \
    $IMAGE \
    bash /scripts/test_artifacts.sh $ARCH $VERSION
```

`test_artifacts.sh`:
```bash
ARCH=$1
VERSION=$2

npm install /node/vectordb-$VERSION.tgz
npm install /node/dist/lancedb-vectordb-linux-$ARCH-gnu-$VERSION.tgz

node -e "const lancedb = require('vectordb'); lancedb.connect(\"./.lancedb\")"
```
</details>

**Comment by wjones127** [comment]
✅ Also tested on x86. I think this is ready to go.

**Pull request merged**